core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -643,7 +645,12 @@ private[spark] object Utils extends Logging {
           fileOverwrite)
       } else {
         val in = fs.open(innerPath)
-        val targetFile = new File(targetDir, innerPath.getName)
+        val targetFile: File =
+          if (!filename.isEmpty) {
+            new File(targetDir, filename)
+          } else {
+            new File(targetDir, innerPath.getName)
+          }
```

Comment:
Yes, the problem here is that now `targetDir` is the parent directory of where `path` should be, and the children are being written directly to that parent path. It needs some code to create this directory corresponding to `path` before downloading the children.